cpp/src/groupby/hash/groupby.cu

```diff
@@ -84,9 +94,114 @@ bool constexpr is_hash_aggregation(aggregation::Kind t)
+                              rmm::device_vector<size_type> const& gather_map,
+                              size_type map_size,
+                              cudaStream_t stream,
+                              rmm::mr::device_memory_resource* mr)
+    : i(col_idx),
+      col(col),
+      sparse_results(sparse_results),
+      dense_results(dense_results),
+      gather_map(gather_map),
+      map_size(map_size),
+      stream(stream),
+      mr(mr)
+  {
+  }
+
+  auto to_dense_agg_result(cudf::aggregation const& agg)
+  {
+    auto s                  = sparse_results.get_result(i, agg);
+    auto dense_result_table = cudf::detail::gather(
+      table_view({s}), gather_map.begin(), gather_map.begin() + map_size, false, mr, stream);
+    return dense_result_table->release()[0];
+  }
+
+  // Enables conversion of ARGMIN/ARGMAX into MIN/MAX
+  auto transformed_result(aggregation::Kind const& agg_kind)
+  {
+    auto transformed_agg = std::make_unique<aggregation>(agg_kind);
+    auto arg_result      = to_dense_agg_result(*transformed_agg);
+    // We make a view of ARG(MIN/MAX) result without a null mask and gather
+    // using this map. The values in data buffer of ARG(MIN/MAX) result
+    // corresponding to null values was initialized to ARG(MIN/MAX)_SENTINEL
+    // which is an out of bounds index value (-1) and causes the gathered
+    // value to be null.
+    column_view null_removed_map(
+      data_type(type_to_id<size_type>()),
+      arg_result->size(),
+      static_cast<void const*>(arg_result->view().template data<size_type>()));
+    auto transformed_result =
+      cudf::detail::gather(table_view({col}),
+                           null_removed_map,
+                           arg_result->nullable() ? cudf::detail::out_of_bounds_policy::IGNORE
+                                                  : cudf::detail::out_of_bounds_policy::NULLIFY,
+                           cudf::detail::negative_index_policy::NOT_ALLOWED,
+                           mr,
+                           stream);
+    return transformed_result->release()[0];
+  };
+
+  // Declare overloads for each kind of aggregation to dispatch
+  void visit(cudf::aggregation& agg) override
+  {
+    dense_results->add_result(i, agg, to_dense_agg_result(agg));
+  }
+
+  void visit(cudf::detail::min_aggregation& agg) override
+  {
+    if (col.type().id() == type_id::STRING)
+      dense_results->add_result(i, agg, transformed_result(aggregation::ARGMIN));
+    else
+      dense_results->add_result(i, agg, to_dense_agg_result(agg));
+  }
+
+  void visit(cudf::detail::max_aggregation& agg) override
+  {
+    if (col.type().id() == type_id::STRING)
+      dense_results->add_result(i, agg, transformed_result(aggregation::ARGMAX));
+    else
+      dense_results->add_result(i, agg, to_dense_agg_result(agg));
+  }
+
+  void visit(cudf::detail::mean_aggregation& agg) override
+  {
+    dense_results->add_result(i, agg, to_dense_agg_result(aggregation{agg.kind}));
```

Comment:
Shouldn't there be a binary op here to divide the sum by the count?

Reply:
updated